examples/connext_dds/xml_application_env/c++11/xml_application_env.cpp

```diff
@@ -0,0 +1,150 @@
+/*******************************************************************************
+ (c) 2020 Copyright, Real-Time Innovations, Inc.  All rights reserved.
+ RTI grants Licensee a license to use, modify, compile, and create derivative
+ works of the Software.  Licensee has the right to distribute object form only
+ for use with RTI products.  The Software is provided "as is", with no warranty
+ of any type, including any warranty for fitness for any purpose. RTI is under
+ no obligation to maintain or support the Software.  RTI shall not be liable for
+ any incidental or consequential damages arising out of the use or inability to
+ use the software.
+ ******************************************************************************/
+
+#include <chrono>
+#include <cstdlib>
+#include <dds/dds.hpp>
+#include <iostream>
+#include <rti/rti.hpp>
+
+using namespace dds::core;
+using namespace dds::core::xtypes;
+using namespace dds::domain;
+using namespace dds::pub;
+using namespace dds::sub;
+using namespace rti::core;
+
+const std::string QOS_URL = "file://application.xml";
+const std::string PARTICIPANT_NAME = "domain_participant_library::participant";
+const std::string WRITER_NAME = "publisher::writer";
+const std::string READER_NAME = "subscriber::reader";
+const std::string KVP_TYPE_NAME = "KeyValuePair";
+
+class MyDataReaderListener : public NoOpDataReaderListener<DynamicData> {
+    virtual void on_data_available(DataReader<DynamicData> &reader)
+    {
+        LoanedSamples<DynamicData> samples = reader.take();
+        for (LoanedSamples<DynamicData>::iterator sampleIt = samples.begin();
+             sampleIt != samples.end();
+             ++sampleIt) {
+            // If the reference we get is valid data, it means we have actual
+            // data available, otherwise we received metadata.
+            if (sampleIt->info().valid()) {
+                std::cout << sampleIt->data() << std::endl;
+            } else {
+                std::cout << "  Received metadata" << std::endl;
+            }
+        }
+    }
+};
+
+
+// Writes environment variables to a DDS topic.  Each environment variable is
+// written as a key/value pair to the DDS bus.
+//
+//  writer - a DataWriter for DnyamicData
+//  kvpType - a key value pair DynamicType
+//  env - the environment array
+void publish_env(int id,
+        DataWriter<DynamicData> writer,
```

Comment:
These parameters should be references to avoid copies:
- `DataWriter<DynamicData>& writer`
- `DynamicType& kvpType`,
...

Reply:
Good call.

Reply:
Good catch.  I'll make that change.